os/pm/pm_timedStay.c

```diff
@@ -0,0 +1,134 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/wdog.h>
+
+#include <pm_timer/pm_timer.h>
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Private Variables
+ ************************************************************************/
+
+/* This array maps the pid to their respective wdog timer. 
+ * Assumption: when a thread does a timed lock , only the same thread 
+ * can unlock the "timed lock" before it expire */
+static WDOG_ID pid_timerMap[CONFIG_MAX_TASKS];
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+static void timer_timeout(int argc, uint32_t state, uint32_t pid)
+{
+	/* PM transition will be relaxed here */
+	if (pid_timerMap[pid] != NULL) {
+		pm_relax(PM_IDLE_DOMAIN, (enum pm_state_e)state);
+		wd_delete(pid_timerMap[pid]);
+        	pid_timerMap[pid] = NULL;
+	}
+}
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: pm_timedStay
+ *
+ * Description:
+ *   This function locks PM state transition for a specific duration.  
+ * 
+ * Parameters:
+ *   state - state to be suspended
+ *   timer_interval - expected lock duration in millisecond
+ *
+ * Return Value:
+ *   0 - success
+ *   -1 - error
+ *
+ ************************************************************************/
+
+int pm_timedStay(enum pm_state_e state, unsigned int timer_interval)
+{
+	int pid = getpid();
+
+	/* Check if there is already a wdog lock timer running for 
+	 * the process */
+	if (pid_timerMap[pid] != NULL) {
```

Comment:
If there is already a timer, it is recommended that we do not clear the timer and start again after wd_cancel.

And if timer_interval is entered as 0, please call wd_cancel and wd_delete only(not start).

Reply:
I would be failed when a new thread with the same pid hash comes in.